src/poller.ts

```diff
@@ -0,0 +1,69 @@
+import type { Options } from './types'
+
+export default class Poller {
+  private timeoutId: number
+  private isRunning: boolean
+
+  constructor(
+    private fetchFlagsFn: () => Promise<any>,
+    private configurations: Options,
+    private pollInterval: number,
+    private maxAttempts: number
+  ) {}
+
+  public start(): void {
+    if (this.isPolling()) {
+      this.logDebug('Polling was requested but is already running - only one poller can run at a time.')
+      return
+    }
+
+    this.isRunning = true
+
+    // Don't start polling immediately as we have alreadu fetched flags on client initialization
+    window.setTimeout(() => this.poll(), this.configurations.pollingInterval)
+  }
+
+  private poll(): void {
+    this.attemptFetch().finally(() => {
+      this.timeoutId = window.setTimeout(() => this.poll(), this.configurations.pollingInterval)
+    })
+  }
+
+  private async attemptFetch(): Promise<void> {
+    for (let i = 1; i <= this.maxAttempts; i++) {
+      const error = await this.fetchFlagsFn()
+
+      if (!error) {
+        this.logDebug('Successfully polled for flag updates')
+        return
+      }
+
+      this.logDebug('Error when polling for flag updates', error)
+
+      // Retry fetching flags
+      if (i < this.maxAttempts) {
+        this.logDebug(`Polling for flags attempt #${i} failed. Remaining attempts: ${this.maxAttempts - i}.`)
+      } else {
+        this.logDebug(`Maximum attempts reached for polling for flags. Next poll in ${this.pollInterval}ms.`)
+      }
+    }
```

Comment:
I'm not sure we can - as we don't return early inside the `if` , so the code outside of it would be executed regardless. I opted for an `if-else` here instead of an early return just for clarity.

Reply:
I've simplified this now inverted the conditions to do an early return

Reply:
The early return happened in the `!error` part. So it would only get outside the loop where the `Max attempts reached` bit is if it completed the loop, which it won't with a successful fetch

Reply:
Oh *facepalm* - your change was placed outside of the loop. My bad.